### PR TITLE
Add staleElementRetry helper

### LINKED ISSUE #748: Better protection against stale element reference (feature request)
It would be nice if webdriverio will try to re-fetch the stale element if such error occurs.
Or would it?

For instance I was having some problems testing button with a counter, where waitForCountToIncrease would fail very often because webdriver would get counter reference, but before next check the count updates and re-renders the whole button, and then webdriver check fails.

Natural place would probably be [waitUntil](https://github.com/webdriverio/webdriverio/blob/ce51baa448714e69f19899718236f988417b4b3d/lib/commands/waitUntil.js#L93), with similar check to timeout but using attempts counter, I guess.

Thanks!

### PR DESCRIPTION
**Pull request by georgecrawford**
Fixes #748

### CODE REVIEW

**Comment by christian-bromann** [comment]
@georgecrawford great contribution, thanks!
I've seen couple of occurrences where the command name is wrong. Could you fix this?

File: lib/commands/addValue.js [hunk 1]
```diff
@@ -67,6 +68,7 @@ module.exports = function addValue (selector, value) {
 
         return this.unify(elementIdValueCommands);
 
-    });
+    })
+    .catch(staleElementRetry.bind(this, 'isVisible', arguments));
```
Comment (christian-bromann) [inline lib/commands/addValue.js]: e.g. here, it should be `.catch(staleElementRetry.bind(this, 'addValue', arguments));`

**Comment by georgecrawford** [comment]
Damn, sorry! Seems I got lazy halfway through.

**Comment by DV3D** [comment]
Great work!! hope this gets into a new release quickly :+1:

**Comment by christian-bromann** [comment]
Let's do it. :+1:

**Comment by georgecrawford** [comment]
@christian-bromann We should discuss what to do with this in v4.0. I could apply the same patch, or we could come up with a nicer solution.

**Comment by christian-bromann** [comment]
@georgecrawford for v4 we should use an `afterCommand` hook if possible. Problem will be that if a staleElementError appears we can't just retry the last command with the same parameters (e.g. `elementIdVisible`) we need to query the element again. I guess a solution could be to use the command history and execute the the command that executed an element command.

**Pull request merged**